includes/class-swedbank-pay-scheduler.php

```diff
@@ -0,0 +1,144 @@
+
+	/**
+	 * Singleton pattern
+	 *
+	 * @return Swedbank_Pay_Scheduler
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Logger instance.
+	 *
+	 * @var WC_Logger
+	 */
+	private $logger;
+
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		$this->logger = \wc_get_logger();
+
+		add_action( 'swedbank_pay_scheduler_run', array( $this, 'run' ), 10, 2 );
+	}
+
+	/**
+	 * Log message.
+	 *
+	 * @param string $message The message to log.
+	 */
+	private function log( $message ) {
+		$this->logger->info( $message, array( 'source' => 'swedbank_pay_queue' ) );
+	}
+
+	/**
+	 * Code to execute for each item in the queue.
+	 *
+	 * @throws \Exception If the webhook data is invalid or if the order cannot be found.
+	 *
+	 * @param string $payment_method_id The payment method ID.
+	 * @param string $webhook_data The webhook data in JSON format.
+	 *
+	 * @return false|null
+	 */
+	public function run( $payment_method_id, $webhook_data ) {
+		$this->log( sprintf( 'Start task: %s', wp_json_encode( array( $payment_method_id, $webhook_data ) ) ) );
+
+		try {
+			$data = json_decode( $webhook_data, true );
+			if ( empty( $data ) ) {
+				throw new \Exception( 'Invalid webhook data' );
+			}
+
+			if ( ! isset( $data['paymentOrder']['id'] ) ) {
+				throw new \Exception( 'Error: Invalid paymentOrder value' );
+			}
+
+			if ( ! isset( $data['transaction']['number'] ) ) {
+				throw new \Exception( 'Error: Invalid transaction number' );
+			}
+
+			$transaction_number = $data['transaction']['number'];
+			$payment_order_id   = $data['paymentOrder']['id'];
+
+			// Get order by `orderReference`.
+			if ( isset( $data['orderReference'] ) ) {
+				$order = wc_get_order( $data['orderReference'] );
+				if ( ! $order ) {
+					throw new \Exception( 'Failed to find order: ' . $data['orderReference'] );
+				}
+
+				$this->log( "[SCHEDULER]: Found order #{$order->get_id()} by order reference {$data['orderReference']}." );
+			} else {
+				// Get Order by Payment Order Id.
+				$order = swedbank_pay_get_order( $payment_order_id );
+				if ( ! $order ) {
+					throw new \Exception( "Failed to find order: $payment_order_id" );
+				}
+
+				$this->log( "[SCHEDULER]: Found order {$order->get_id()} by payment Order ID $payment_method_id." );
+			}
+
+			$gateway = swedbank_pay_get_payment_method( $order );
+			if ( ! $gateway ) {
+				throw new \Exception( "Cannot retrieve payment gateway instance: $payment_method_id" );
+			}
+
+			if ( ! property_exists( $gateway, 'api' ) ||
+				! in_array( $order->get_payment_method(), Swedbank_Pay_Plugin::PAYMENT_METHODS, true ) ) {
+				$this->log( "[ERROR]: Order {$order->get_id()} has not been paid with the swedbank pay. Payment method: {$order->get_payment_method()}" );
+
+				return false;
+			}
+
+			$transactions = (array) $order->get_meta( '_swedbank_pay_transactions' );
```

Comment:
Question: Just for my own knowledge, and to make sure we don't miss anything. `WC_Order::get_meta` has the parameter `$single`, which is defaulted to true. If this is an array of transactions, is it something we need to get as not a single instead? I just want to make sure we don't end up with something that works when testing but not in production for edge cases.

Reply:
This is the code from the existing background queue processor, so it should work in production. 

From what I can see, we're only storing a single array [here](https://github.com/krokedil/swedbank-pay-woocommerce-paymentmenu/blob/main/includes/class-swedbank-pay-api.php#L673). In this case, `$single = true` should return this entire array.